File: apps/web/__tests__/integration/handlers/graph.ts
import { rest } from 'msw'

export const graphHandlers = [
  rest.get('/api/graph', (req, res, ctx) => {
    return res(
      ctx.json({ nodes: [{ id: 'node-1', label: 'AI', level: 'current' }], edges: [] })
    )
  }),
  rest.post('/api/graph/expand', (req, res, ctx) => {
    return res(
      ctx.json({ nodes: [{ id: 'node-2', label: 'Testing', level: 'child' }], edges: [{ source: 'node-1', target: 'node-2', weight: 1 }] })
    )
  }),
  rest.get('/api/graph/nodes/:id/resources', (req, res, ctx) => {
    return res(
      ctx.json({ items: [{ id: 'uuid-1', title: 'Mock Resource', status: 'READY' }] })
    )
  }),
]

File: apps/api/schemas/graph.py
"""Pydantic schemas for graph API endpoints."""

from pydantic import BaseModel, field_validator


class GraphNode(BaseModel):
    """A node in the knowledge graph."""

    id: str
    label: str
    level: str  # "root", "current", "child", "parent"
    node_type: str  # "root", "category", "topic"
    resource_count: int = 0  # Number of resources associated with this node


class GraphEdge(BaseModel):
    """An edge between two nodes in the knowledge graph."""

    source: str
    target: str
    weight: int


class GraphResponse(BaseModel):
    """Response schema for graph data."""

    nodes: list[GraphNode]
    edges: list[GraphEdge]


class GraphExpandRequest(BaseModel):
    """Request schema for expanding a graph node."""

    node_id: str
    direction: str = "out"  # "out", "in", "both"

    @field_validator("direction")
    @classmethod
    def validate_direction(cls, v: str) -> str:
        if v not in ("out", "in", "both"):
            raise ValueError("direction must be one of: out, in, both")
        return v
